src/js/letters/components/Address.jsx

```diff
@@ -0,0 +1,151 @@
+import React from 'react';
+import _ from 'lodash';
+import { set } from 'lodash/fp';
+
+import ErrorableSelect from './ErrorableSelect';
+import ErrorableTextInput from './ErrorableTextInput';
+import { isNotBlank, isBlankAddress, isValidUSZipCode, isValidCanPostalCode } from '../../common/utils/validations';
+import { countries, states } from '../../common/utils/options-for-select';
+
+/**
+ * Input component for an address.
+ *
+ * No validation is provided through a currently stubbed isAddressValid function.
+ */
+class Address extends React.Component {
+  constructor() {
+    super();
+    this.handleChange = this.handleChange.bind(this);
+    this.validateAddressField = this.validateAddressField.bind(this);
+    this.validatePostalCode = this.validatePostalCode.bind(this);
+  }
+
+  componentWillMount() {
+    this.id = _.uniqueId('address-input-');
+  }
+
+  // TODO: Look into if this is the best way to update address,
+  // it is incredibly slow right now
+  handleChange(path, update) {
+    let address = set(path, update, this.props.value);
+    // if country is changing we should clear the state
+    if (path === 'country') {
+      address = set('state', '', address);
+    }
+
+    this.props.onUserInput(address);
+  }
+
+  validateAddressField(field) {
+    if (this.props.required || !isBlankAddress(this.props.value)) {
```

Comment:
Excuse my ignorance - what is `this.props.required` as it relates to an `Address` component?

Reply:
It's meant to represent the question of whether the address as a whole is required. If it's not required, we may not care that you submit a valid and complete address. If it is required, we do care that the address is valid and complete. A valid address needs to have at least a street, city, state, zip, and country. So we can use the general sense of whether or not the address is required to require each of those individual fields.